src/NHibernate/Hql/Ast/ANTLR/HqlSqlWalker.cs

```diff
@@ -1072,7 +1088,10 @@ IASTNode GenerateNamedParameter(IASTNode delimiterNode, IASTNode nameNode)
 				// Add the parameter type information so that we are able to calculate functions return types
 				// when the parameter is used as an argument.
 				if (namedParameter.IsGuessedType)
+				{
+					_guessedParameterTypes[paramSpec] = namedParameter.Type;
 					parameter.GuessedType = namedParameter.Type;
```

Comment:
can we do here this?

```csharp
paramSpec.ExpectedType = namedParameter.Type;
```

Reply:
Or is it too early?

Reply:
Yes. Too early. HQL parameter detection logic is not yet executed

Reply:
I have an idea how to get rid of `_guessedParameterTypes`. But it requires new method in `IQuery`, so not a good fit for a patch release.